### PR TITLE
DOC: Change extract-text.md example codes from using cm to tm

### LINKED ISSUE #2431: DOC: Example code doesn't give the right output (fix + proven, didn't want to create a pull request for that)
I was trying to use the exact same example mentioned in [here](https://pypdf.readthedocs.io/en/latest/user/extract-text.html#example-1-ignore-header-and-footer), but it gives blank output, even though I copied the same code, and same [PDF file](https://github.com/py-pdf/pypdf/blob/main/resources/GeoBase_NHNC1_Data_Model_UML_EN.pdf). (Fix is at the bottom of this issue report)

## Environment

Debian

```bash
$ python -m platform
Linux-6.1.0-12-amd64-x86_64-with-glibc2.36

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==4.0.1, crypt_provider=('cryptography', '41.0.7'), PIL=10.2.0
```

## Code + PDF

This is a minimal, complete example that shows the issue (same example from documentation):

```python
from pypdf import PdfReader

reader = PdfReader("GeoBase_NHNC1_Data_Model_UML_EN.pdf")
page = reader.pages[3]

parts = []


def visitor_body(text, cm, tm, font_dict, font_size):
    y = cm[5]
    if y > 50 and y < 720:
        parts.append(text)


page.extract_text(visitor_text=visitor_body)
text_body = "".join(parts)

print(text_body)
```

## Fix

Just change `cm` to `tm`. The selection of height must be from the text matrix, not current matrix.

[Here](https://github.com/py-pdf/pypdf/blob/main/resources/GeoBase_NHNC1_Data_Model_UML_EN.pdf)'s to the PDF file.

### PR DESCRIPTION
**Pull request by etern4l-white**
## TL;DR
Fixes [#2431](https://github.com/py-pdf/pypdf/issues/2431)
Changed `cm` (current_matrix) to `tm` (text matrix). 

## Problem

In the extract-text documentation [here](https://github.com/py-pdf/pypdf/blob/main/docs/user/extract-text.md#example-1-ignore-header-and-footer), the example codes that are used won't produce the correct output. 

For example, the first code snippet should output the text of table of contents, but it outputs nothing. The second code snippet is supposed to convert page 4 from the PDF to a SVG, including the text, but it only outputs empty fields.

## Reason

The coordination process should be using the text matrix instead of current matrix.

## Solution/update

Just changed the `cm` to `tm` in the code snippets.

### CODE REVIEW

**Comment by etern4l-white** [comment]
This is my first pull request ever. Any comments?

File: docs/user/extract-text.md [hunk 1]
```diff
@@ -119,7 +119,7 @@ def visitor_svg_rect(op, args, cm, tm):
 
 
 def visitor_svg_text(text, cm, tm, fontDict, fontSize):
```
Comment (stefan6419846) [inline docs/user/extract-text.md:121]:
```suggestion
def visitor_svg_text(text, cm, tm, font_dict, font_size):
```

Although unrelated, we should probably change this to be consistent with the first example (and follow the naming from PEP8).

**Comment by pubpub-zz** [comment]
I personally do not recommend to extract text using cm : We've observed many case where the actual text and ordering is not valid.
Maybe you should try to use the new text extraction with the layout.

**Comment by stefan6419846** [comment]
@pubpub-zz This is an example inside our docs which somehow stopped working correctly, maybe because due to internal changes, maybe because it has not worked at all in the past. If ever, we should explain the new layout mode besides the "classic" mode in the docs. But this can be part of another PR - our examples should work and yield correct results to avoid frustrations.

**Comment by etern4l-white** [comment]
> @pubpub-zz This is an example inside our docs which somehow stopped working correctly, maybe because due to internal changes, maybe because it has not worked at all in the past. If ever, we should explain the new layout mode besides the "classic" mode in the docs. But this can be part of another PR - our examples should work and yield correct results to avoid frustrations.

By the way, in pypdf version 3 (i guess) documentation it uses `tm` intead of `cm` , too. [See here](https://github.com/py-pdf/pypdf/blob/3.0.0/docs/user/extract-text.md#example-1-ignore-header-and-footer). I don't know the exact underlying reason.

**Comment by stefan6419846** [comment]
This has been a change by @pubpub-zz: https://github.com/py-pdf/pypdf/commit/bcd85c4e4b46116a049fbba54b5b0aa7260ff041

**Comment by etern4l-white** [comment]
Hm, so `tm` was making issues, and that change fixed the issue? That makes sense, but I believe the example in the docs should be working, too. I'm kind of confused to be honest, since I'm not that deep in the internals of how it works. 

Maybe the example PDF was the problem? Because if that change was to solve a problem, then maybe `tm` was producing an issue in most of PDFs, but not in the example PDF, so switching to `cm` might have fixed the issue for alot of PDFs, but not the one in the example. 

I think for me to validate this assumption I should test it on many PDFs. I will do later today because I got an assignment to do 😅.

**Comment by stefan6419846** [comment]
Did you have a chance to further check this already?

**Comment by MartinThoma** [comment]
Hey @etern4l-white  are there any updates? :innocent:

**Comment by stefan6419846** [comment]
Should fix #2881 as well.

**Comment by stefan6419846** [comment]
Does this have any effect on the text size? According to the quote in https://github.com/py-pdf/pypdf/issues/2881#issuecomment-2380648082, the text size is somehow affected the multiplied matrix.

**Comment by r-kamei** [comment]
I tested the current examples, and `cm` alone still does not produce the expected output.

Since the documentation already recommends using `txt2user = mult(tm, cm)`, would it make sense to update the examples to use that transformation instead of `cm` or `tm` alone?

The text-size question seems separate from this coordinate issue.

```python
    from pypdf import mult
    ...
    def visitor_body(text, cm, tm, font_dict, font_size):
        txt2user = mult(tm, cm)
        y = txt2user[5]
        if 50 < y < 720 or y == 0:
            parts.append(text)
    ...
    def visitor_svg_text(text, cm, tm, font_dict, font_size):
        txt2user = mult(tm, cm)
        (x, y) = (txt2user[4], txt2user[5])
        dwg.add(dwg.text(text, insert=(x, y), fill="blue"))
```

**Comment by stefan6419846** [comment]
> Since the documentation already recommends using `txt2user = mult(tm, cm)`, would it make sense to update the examples to use that transformation instead of `cm` or `tm` alone?

I am open to merging a corresponding PR, as this is what never has been changed in this PR.

**Comment by r-kamei** [comment]
Thanks for the clarification.

I've opened a follow-up PR that updates the visitor examples to use `mult(tm, cm)` instead of relying on `cm` or `tm` alone:

#4066
